design/src/rfcs/rfc0016_flexible_checksum_support.md

```diff
@@ -0,0 +1,1442 @@
+    /// fixed length.
+    pub chunk_length: Option<usize>,
+    /// The length of each trailer sent within an `AwsChunkedBody`. Necessary in
+    /// order to correctly calculate the total size of the body accurately.
+    pub trailer_lens: Vec<usize>,
+}
+
+impl AwsChunkedBodyOptions {
+    /// Create a new [`AwsChunkedBodyOptions`][AwsChunkedBodyOptions]
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Set stream length
+    pub fn with_stream_length(mut self, stream_length: usize) -> Self {
+        self.stream_length = Some(stream_length);
+        self
+    }
+
+    /// Set chunk length
+    pub fn with_chunk_length(mut self, chunk_length: usize) -> Self {
+        self.chunk_length = Some(chunk_length);
+        self
+    }
+
+    /// Set a trailer len
+    pub fn with_trailer_len(mut self, trailer_len: usize) -> Self {
+        self.trailer_lens.push(trailer_len);
+        self
+    }
+}
+
+#[derive(Debug, PartialEq, Eq)]
+enum AwsChunkedBodyState {
+    WritingChunkSize,
+    WritingChunk,
+    WritingTrailers,
+    Closed,
+}
+
+/// A request body compatible with `Content-Encoding: aws-chunked`
+///
+/// Chunked-Body grammar is defined in [ABNF] as:
+///
+/// ```txt
+/// Chunked-Body    = *chunk
+///                   last-chunk
+///                   chunked-trailer
+///                   CRLF
+///
+/// chunk           = chunk-size CRLF chunk-data CRLF
+/// chunk-size      = 1*HEXDIG
+/// last-chunk      = 1*("0") CRLF
+/// chunked-trailer = *( entity-header CRLF )
+/// entity-header   = field-name ":" OWS field-value OWS
+/// ```
+/// For more info on what the abbreviations mean, see https://datatracker.ietf.org/doc/html/rfc7230#section-1.2
+///
+/// [ABNF]:https://en.wikipedia.org/wiki/Augmented_Backus%E2%80%93Naur_form
+#[derive(Debug)]
+#[pin_project]
+pub struct AwsChunkedBody<InnerBody> {
+    #[pin]
+    inner: InnerBody,
+    #[pin]
+    state: AwsChunkedBodyState,
+    options: AwsChunkedBodyOptions,
+}
+
+// Currently, we only use this in terms of a streaming request body with checksum trailers
+type Inner = ChecksumBody<SdkBody>;
+
+impl AwsChunkedBody<Inner> {
+    /// Wrap the given body in an outer body compatible with `Content-Encoding: aws-chunked`
+    pub fn new(body: Inner, options: AwsChunkedBodyOptions) -> Self {
+        Self {
+            inner: body,
+            state: AwsChunkedBodyState::WritingChunkSize,
+            options,
+        }
+    }
+
+    fn encoded_length(&self) -> Option<usize> {
+        if self.options.chunk_length.is_none() && self.options.stream_length.is_none() {
+            return None;
+        }
+
+        let mut length = 0;
+        let stream_length = self.options.stream_length.unwrap_or_default();
+        if stream_length != 0 {
+            if let Some(chunk_length) = self.options.chunk_length {
+                // I don't think we'll hit this case b/c we only ever send things in one chunk
+                assert!(chunk_length > MINIMUM_CHUNK_LENGTH);
```

Comment:
I carried over this assert when I implemented this part by reading the Go v2 SDK. Looking at the referenced Sigv4, it seems to be a bit unclear on whether or not there are limits. All this stuff works right now because we don't use the chunking at all. I brought it over in case we want to implement signed chunks. I'll write a test to ensure the chunking works.

Reply:
I removed this check because it's not  necessary